### ISSUE TITLE
`Process.exposed_outputs` does not work with nested namespaces

### ISSUE THREAD

**Issue by sphuber** [aiidateam/aiida-core#3533]
Simple test case to demonstrate the problem:
```
class ChildProcess(Process):

    _node_class = orm.WorkflowNode

    @classmethod
    def define(cls, spec):
        super(ChildProcess, cls).define(spec)
        spec.input('input', valid_type=orm.Int)
        spec.output('output', valid_type=orm.Int)
        spec.output('name.space', valid_type=orm.Int)

    @override
    def run(self):
        self.out('output', self.inputs.input)
        self.out('name.space', self.inputs.input)


class ParentProcess(Process):

    _node_class = orm.WorkflowNode

    @classmethod
    def define(cls, spec):
        super(ParentProcess, cls).define(spec)
        spec.input('input', valid_type=orm.Int)
        spec.expose_outputs(ChildProcess)

    @override
    def run(self):
        from aiida.manage.manager import get_manager
        runner = get_manager().create_runner()
        results, node = runner.run_get_node(ChildProcess, **self.inputs)
        self.out_many(self.exposed_outputs(node, ChildProcess))


class TestProcessNamespace(AiidaTestCase):

    def test_namespaced_process(self):
        results, node = run_get_node(ParentProcess, input=orm.Int(5))
        print(results)
```
The results will only contain `output` and not the `name.space` output.

**Comment by JPchico** [comment]
An issue that might be related to this, is that when having nested workchains, even when does not specify any `exposed_outputs` they are still tried to be written to the `outputs` node, resulting in the workchain exiting in an abnormal fashion.

**Comment by espenfl** [comment]
@sphuber What is the status of this?

**Comment by espenfl** [comment]
Any updates to this?

**Comment by JPchico** [comment]
@sphuber and @espenfl I have been looking at this again, since I have a workflow, in which I relax several structures, perform several operations with them and then expose the outputs of of final relaxations of each one of them, in principle I want to attach each one of them in their own namespaces. 

I can somewhat get around the issue by using the following trick
```python
spec.expose_outputs(cls._workchain, namespace='name')
spec.outputs.dynamic = True
...
self.out_many({'name__' + key: self.ctx.name_workchain.outputs[key] for key in self.ctx.name_workchain.outputs})
```

the workchain ends and all outputs are present, but the calculation ends with the state `Finished [11] The process did not register a required output.`

My guess is that it is related to this in `aiida-core/aiida/engine/processes/process.py`
```python
for top_name, port_namespace in output_key_map.items():
    # collect all outputs belonging to the given top_name
    for port_name in top_namespace_map[top_name]:
        if port_namespace is None:
            result[port_name] = process_outputs_dict[port_name]
        else:
            result[port_namespace + namespace_separator + port_name] = process_outputs_dict[port_name]
```

I guess that the issue is that when looping over the `top_namespace_map[top_name]` the `port_name` either does not detect or is not able to pass nested namespaces to the dictionary.

**Comment by espenfl** [comment]
Thanks @JPchico for bringing this alive again. I think it would be awesome to try to enable this functionality as it requires boilerplate on the plugin side and also seems odd when we have it on the input side.

**Comment by bosonie** [comment]
I take the occasion to also bring the attention to a related issue.
Since many plugins started to use the `BaseRestartWorkChain`, when this is fixed, we should also fix  #4623

**Comment by JPchico** [comment]
I think I actually know what is the issue, what is happening is that when one has a nested namespace the separator is no longer `.` if not it is `__` this results in that in `top_namespace_map` one does not store a `{top_name : name}` pair instead one saves a `{top_name: top_name}` pair. 

```python
for port_name in process_outputs_dict:
    top_namespace = port_name.split(namespace_separator)[0]
    top_namespace_map[top_namespace].append(port_name)
```
The question is whether one should have a conditional to check if `__` is present in the key and it should be separated as such, or if the links themselves should be stored as `top.name` instead of `top__name`.

**Comment by JPchico** [comment]
Changing the generation of `process_outputs_dict` to take into consideration this and replace the `__` by `.` can lead to one solving this issue. 
```python
process_outputs_dict = {(entry.link_label if '__' not in entry.link_label else entry.link_label.replace('__','.')): entry.node for entry in node.get_outgoing(link_type=link_types)}
```
Of course this would mean that if an user for some reason has an output with name `__` it would be taken as a dictionary namespace instead of a flat name, which I guess it is acceptable. 

@sphuber  if you are happy with this solution I can make a PR with this fix.

**Comment by sphuber** [comment]
Thanks for reviving this issue and pinpointing the problem. You correctly analysed the problem, but there is a better solution. I didn't see proposed to open a PR, and came up with the solution and test in the meantime( see PR #4863). Please have a look and test if that fixes the problem for you.

**Comment by JPchico** [comment]
@sphuber fantastic! I did not know about the `.nested()` method that is indeed much more elegant. My small tests indeed seem to show that this would work. I was just waiting to discuss the solution before opening a PR, because I imagined that there was a smarter way of doing it than what I did. 

I guess that this does not solve #4623 as one would need to change 
```python
        for name, port in self.spec().outputs.items():

            try:
                output = node.get_outgoing(link_label_filter=name).one().node
```
to be more akin to what is in `aiida-core/aiida/engine/processes/process.py` as of that way one can loop over the nested namespaces.

**Comment by bosonie** [comment]
> I guess that this does not solve #4623 as one would need to change
> 
> ```python
>         for name, port in self.spec().outputs.items():
> 
>             try:
>                 output = node.get_outgoing(link_label_filter=name).one().node
> ```
> 
> to be more akin to what is in `aiida-core/aiida/engine/processes/process.py` as of that way one can loop over the nested namespaces.

Correct

**Comment by sphuber** [comment]
Although the fix in the PR I pushed fixes the common problem and the test case I added for this, it breaks another test that uses exposed inputs and outputs. I have been breaking my head over this for a bit now but I am not even sure that what the test is testing is correct. The difficulty comes due to the optional `agglomerate` argument in the `exposed_inputs` and `exposed_outputs` functions. I originally came up with this keyword for the inputs, in order to simplify the life of users of heavily nested workchains. I don't think it is actually being used but it complicates the logic enormously. So it will take some more time to finalize the PR and the fix I'm afraid.

**Comment by JPchico** [comment]
Hmmm yes I see the issue, I guess that when one is calling `_get_namespace_list` is that the issue pops up right? I'm a bit confused on what is actually breaking it, since the change in exposed outputs, should only affect the `AttirbuteDict` that one gets after the calculation.

**Comment by sphuber** [comment]
Looked a bit more at it and have fixed the issue. Tests are passing now. If the people in this thread who had the original problem can confirm the patch fixes the issue we can have it reviewed and merged.

**Comment by JPchico** [comment]
@sphuber I can confirm that the fix solves the issue. Thank you very much!

**Comment by espenfl** [comment]
@sphuber looks good, confirmed. Thanks a lot.

**Comment by JPchico** [comment]
Hello @sphuber is there anything else that need to be done to be able to push this? I have also an idea for #4623, it is not very elegant, but it seems to work with a simple test. I could make a PR for it if this solution is deemed acceptable or we can discuss other options on how to solve it.

**Comment by sphuber** [comment]
@JPchico no, I have opened the PR and have notified other developers. All it needs is a formal review from them (they know you and others have confirmed it fixes their problem) and then we can merge. I am pushing to get this released in `1.6.2` a.s.a.p. before we move the `develop` branch to the `v2.0.0` state.
